[RAR] Don't do I/O on SDK-provided references

ladipro opened the pull request:
Fixes #8634

### Context

For SDK/workload references, the SDK is currently already passing relevant metadata such as `AssemblyVersion` and `PublicKeyToken`, so there is no need for RAR to open these files and parse their .NET metadata tables to get this information. Also, barring corrupted installs, these files are guaranteed to exist on disk so there is no need for RAR to do any I/O on them (file-exists, get-time-stamp, ..)

### Changes Made

Tweaked RAR to trust the item metadata passed by the SDK. The feature is gated on the presence of the `FrameworkReferenceName` metadatum which is not documented and is used only by the relevant SDK tasks, to my best knowledge.

SDK does not specify the `Culture` component of assembly names so it's assumed to be `neutral`. This has passed all relevant validation so far. If non-neutral assemblies are ever used as references, we'll need to define a metadatum for SDK to set.

### Testing

- Existing and new unit tests.
- Experimental insertion with an assert that assembly names calculated from item metadata are equivalent to those extracted from assembly files (verifies that the `Culture=neutral` assumption is correct).
- Checked all assemblies specified in all `FrameworkList.xml` files shipped in the SDK (+ workloads) and verified that all of them are culture neutral.

### Perf results

RAR micro-benchmark where the task is invoked with parameters corresponding to building a simple AspNetCore app:
- 13.13 ms -> 3.27 ms per invocation without StateFile cache.
- 15.03 ms -> 5.13 ms per invocation with StateFile cache.

RAR task duration as reported with `/clp:PerformanceSummary` when building a simple AspNetCore app with MSBuild server enabled:
- 20 ms -> 10 ms.

### Notes

- The change is behind a 17.8 changewave.
- No changes have been made to per-project caching (via the `StateFile` parameter) to reduce scope. Changes to the per-project cache file will be done in another PR.

Forgind commented on src/Tasks/AssemblyDependency/AssemblyResolution.cs:
```diff
@@ -48,6 +49,7 @@ internal static class AssemblyResolution
             string sdkName,
             string rawFileNameCandidate,
             bool isPrimaryProjectReference,
+            bool isImmutableFrameworkReference,
```
We're currently saying this is only for immutable framework references, but do you think we might extend that later? If so, maybe we should use a somewhat more generic name?

Forgind replied:
Fair...fullyUnderstoodReferences? Better name with that intent?

ladipro replied:
> I'd say what's interesting is not the immutability, but that we can assume its dependency closure is provided by the framework so don't need to explore it--right?

Not quite right.  The property of "dependency closure is provided" is signaled with the `ExternallyResolved` metadatum. It is a pre-existing behavior that for such assemblies we skip the dependency walk. It is happening here:
https://github.com/dotnet/msbuild/blob/768cad16718650f120dd42acfcc9cf5c4dec555f/src/Tasks/AssemblyDependency/ReferenceTable.cs#L1692-L1699

This new flag means "the file is guaranteed to exist and never change" and in the `Resolver` hierarchy it's used only in `RawFilenameResolver` to skip the file existence check.

As for the name, I was on the fence between "immutable" and "framework"  so I used both 😀

Forgind commented on src/Tasks/AssemblyDependency/ReferenceTable.cs:
```diff
@@ -823,6 +828,29 @@ private static void TryGetAssemblyNameComponent(string fusionName, string compon
+                return null;
+            }
+
+            string name = Path.GetFileNameWithoutExtension(item.ItemSpec);
```
Think we should have a check for if ItemSpec is empty? (Can ItemSpec be empty?)

ladipro replied:
The code has been updated to obtain the name from metadata.

Forgind commented on src/Tasks/AssemblyDependency/ReferenceTable.cs:
```diff
@@ -1225,6 +1253,17 @@ internal static void CalculateParentAssemblyDirectories(List<string> parentRefer
+            if (ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave17_7))
+            {
+                // For a path to be an immutable reference, it must be externally resolved and has a FrameworkReferenceName defined.
+                if (assemblyName == null && !string.IsNullOrEmpty(rawFileNameCandidate) && reference.IsPrimary && reference.ExternallyResolved)
```
Why does it have to be Primary? Why does assemblyName have to be null?

ladipro replied:
We're basically saying: If we are given a file path and not assembly name (`assemblyName` is not set, `rawFileNameCandidate` is set), and it is a primary (as in passed to RAR in a parameter, not discovered in dependency walk) and externally resolved (doesn't require dependency walk) dependency, we can consider it to be immutable.

I believe `ExternallyResolved` implies `IsPrimary` so it could be removed but I think this way it captures the concept better. In fact, we could remove this condition altogether if we trust nobody but the SDK task sets `FrameworkReferenceName`. It's more like defense-in-depth so the new logic kicks in only in scenarios where we tested it.

Forgind replied:
> passed to RAR in a parameter, not discovered in dependency walk

This is the part I find most confusing. Why should it matter how a reference got to RAR? Either it's immutable or it isn't, but having RAR specifically reference it shouldn't affect whether it's the sort of reference that changes. I can imagine a design in which the SDK passes all of its references to RAR (and since this change was quite impactful, I'd assume that's true, or we at least pretend it's true), but I could also imagine a design in which the SDK passes a few entrypoint references to RAR, and the rest comes indirectly even though they're all immutable and even in fixed positions relative to each other. Why do we care about which design is used?

ladipro replied:
I think the design with SDK passing all SDK assembly references is something that's unlikely to change. It's an important optimization and we would take several steps back perf-wise if we had to crack open SDK assemblies to find the closure.

Conceptually I agree that we're artificially limiting the scenario to the specifics of how the system works today. I see it more like defense-in-depth to avoid accidentally exposing something that we would later regret supporting.

Forgind commented on src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs:
```diff
@@ -2295,13 +2295,34 @@ private List<DependentAssembly> GetAssemblyRemappingsFromAppConfig()
                     _cache.SetInstalledAssemblyInformation(installedAssemblyTableInfo);
 
                     // Cache delegates.
-                    getAssemblyName = _cache.CacheDelegate(getAssemblyName);
```
nit:
Any reason to not remove _cache.SetGetLastWriteTime above?

ladipro replied:
Nice catch! Done.

Forgind reviewed (approved):
I'm really excited to see this in action!

rainersigwald commented on src/Tasks.UnitTests/AssemblyDependency/Miscellaneous.cs:
```diff
@@ -8564,5 +8566,56 @@ private static void GenerateRedistAndProfileXmlLocations(string fullRedistConten
+        {
+            InitializeRARwithMockEngine(_output, out MockEngine mockEngine, out ResolveAssemblyReference rar);
+
+            string refPath = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString());
```
I'm a bit surprised by this since the implementation elsewhere assumes that the filename is the assembly name--having a test case where that's not true doesn't break anything?

ladipro replied:
I have updated the test to use the explicit `AssemblyName` metadatum. Not completely sure I understand the concern. If MSBuild is given the right metadata to activate the "no I/O path", it just trusts what we pass it. In the first version of the PR the assembly name was inferred from file name but, as you pointed out, because it's fragile, I have added a check for `AssemblyName` and made the SDK task pass it to us.

Forgind commented on src/Framework/ChangeWaves.cs:
```diff
@@ -27,7 +27,8 @@ internal class ChangeWaves
         internal static readonly Version Wave17_6 = new Version(17, 6);
-        internal static readonly Version[] AllWaves = { Wave17_2, Wave17_4, Wave17_6 };
+        internal static readonly Version Wave17_7 = new Version(17, 7);
+        internal static readonly Version[] AllWaves = { Wave17_2, Wave17_4, Wave17_6, Wave17_7 };
```
Agreed (but in a separate PR)

ladipro replied:
There are no changes gated on 17.2 so I've removed it from the list in this PR.

ladipro commented on src/Framework/ChangeWaves.cs:
```diff
@@ -27,7 +27,8 @@ internal class ChangeWaves
         internal static readonly Version Wave17_4 = new Version(17, 4);
         internal static readonly Version Wave17_6 = new Version(17, 6);
-        internal static readonly Version[] AllWaves = { Wave17_2, Wave17_4, Wave17_6 };
+        internal static readonly Version Wave17_7 = new Version(17, 7);
```
Thank you, I've updated it to 17.8.

ladipro commented on src/Tasks/AssemblyDependency/ReferenceTable.cs:
```diff
@@ -823,6 +828,29 @@ private static void TryGetAssemblyNameComponent(string fusionName, string compon
+                return null;
+            }
+
+            string name = Path.GetFileNameWithoutExtension(item.ItemSpec);
```
Done.

ladipro commented on src/Tasks/AssemblyDependency/ReferenceTable.cs:
```diff
@@ -823,6 +828,29 @@ private static void TryGetAssemblyNameComponent(string fusionName, string compon
+
+            string name = Path.GetFileNameWithoutExtension(item.ItemSpec);
+
+            AssemblyName assemblyName = new AssemblyName($"{name}, Version={version}, Culture=neutral, PublicKeyToken={publicKeyToken}");
```
Although there is a code path where the `Path` attribute is used and assembly name may be different.

Ok, I don't see any existing framework assembly for which filename wouldn't match assembly name but let's future proof this and add an explicit `AssemblyName` to the SDK task: https://github.com/dotnet/sdk/pull/32213

ladipro commented:
I have addressed feedback and run another experimental insertion. It flagged several major improvements in, quite surprisingly, the add-new-project scenario (-30% reference set, for example).

@rokonec noticed that we still open SDK assemblies to read their metadata version ("runtime version"), which doesn't look necessary. I made `GetAssemblyRuntimeVersion` return the hardcoded `v4.0.30319` to remove this I/O, with the following justification:
- None of the assemblies currently distributed with the SDK/workloads appear to have anything else than this canonical version.
- Daniel is not aware of an issue with this assumption.
- The metadata parsing code has a bug that prevents it from detecting WinRT/winmd files, which would have been the only thing potentially broken by the change (#8731). I.e. _it doesn't work anyways_.

@Forgind @rainersigwald, please review. Merging this PR will unblock further RAR perf work.
